server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java

```diff
@@ -163,16 +165,15 @@ public boolean upgradeNode(String downloadJSON) {
     @Override
     public TestSession getNewSession(Map<String, Object> requestedCapability) {
         LOGGER.entering();
-        TestSession session;
+        //verification should be before synchronized block also to avoid unnecessarily acquiring lock
+        if (totalSessionsStarted >= maxSessionsAllowed || forceShutDown) {
+            return logSessionInfo();
+        }
         synchronized (this) {
             if (totalSessionsStarted >= maxSessionsAllowed || forceShutDown) {
```

Comment:
@sebady Yeah I did. Lets assume totalSessionStarted is 24 and maxSessionAllowed is 25. If two threads enter into getNewSession() both of them pass first if condition. Thread1 acquire the lock, get the new session and increase the count totalSessionStarted to 25, release the lock. Now thread2 enter into synchronized block and get into second if condition. This is rare scenario, but we need this check here to avoid these kind of scenario.